### ISSUE TITLE
Patching of arrays can fail

### ISSUE THREAD

**Issue by michaelsauter** [opendevstack/tailor#92]
If there is an array with e.g. 3 entries in both platform and template, and then first two are deleted in the template, the patch fails with `Error from server: Unable to remove invalid index: 2`.

This seems to be an issue at least with rolebindings - it looks like something fishy happens with OpenShift there.

**Comment by henrjk** [comment]
I have observed another issue just now where I am adding environment variables to a deployment config. The initial debug log out put is:

```
Patching dc/be-component ... --> oc patch dc/be-component --type=json --patch [{"op":"add","path":"/spec/template/spec/containers/0/env/10","value":{"name":"A_CLIENT_SECRET","valueFrom":{"secretKeyRef":{"key":"clientSecret","name":"a"}}}},{"op":"add","path":"/spec/template/spec/containers/0/env/11","value":{"name":"SESSION_SECRET","valueFrom":{"secretKeyRef":{"key":"sessionSecret","name":"a"}}}},{"op":"add","path":"/spec/template/spec/containers/0/env/9","value":{"name":"A_CLIENT_ID","valueFrom":{"secretKeyRef":{"key":"clientId","name":"a"}}}},{"op":"add","path":"/spec/template/spec/containers/0/envFrom/1","value":{"configMapRef":{"name":"a"}}}] --namespace=some-wip
failed
2019/11/12 15:48:38 Update aborted: Error from server: Unable to access invalid index: 10
```

When I go into the web console and add A_CLIENT_SECRET manually and then redo the tailor command the patch succeeds, but looks slightly different also:

```
Patching dc/be-component ... --> oc patch dc/be-component --type=json --patch [{"op":"add","path":"/spec/template/spec/containers/0/env/10","value":{"name":"A_CLIENT_SECRET","valueFrom":{"secretKeyRef":{"key":"clientSecret","name":"a"}}}},{"op":"add","path":"/spec/template/spec/containers/0/env/11","value":{"name":"SESSION_SECRET","valueFrom":{"secretKeyRef":{"key":"sessionSecret","name":"a"}}}},{"op":"add","path":"/spec/template/spec/containers/0/envFrom/1","value":{"configMapRef":{"name":"a"}}},{"op":"remove","path":"/spec/triggers/0/imageChangeParams"},{"op":"replace","path":"/spec/triggers/0/type","value":"ConfigChange"},{"op":"add","path":"/spec/triggers/1/imageChangeParams","value":{"automatic":true,"containerNames":["be-component"],"from":{"kind":"ImageStreamTag","name":"be-component:latest","namespace":"some-wip"}}},{"op":"replace","path":"/spec/triggers/1/type","value":"ImageChange"}] --namespace=some-wip
```

tailor version is
0.10.3

**Comment by michaelsauter** [comment]
Hmm interesting case. In your first invocation, it looks like it tries to add path `/spec/template/spec/containers/0/env/10` before adding path `/spec/template/spec/containers/0/env/9`. That is weird and might be the cause of this problem.

BTW - did you mean you added `A_CLIENT_ID` manually? To me it looks like that because that would be index 9 ...

**Comment by henrjk** [comment]
Yes exactly: I added A_CLIENT_ID manually and then it worked. It was sort of the first environment variable in file order that was new.

**Comment by michaelsauter** [comment]
I have verified that this is due to multiple patches that cannot be resolved in the order that Tailor tries to apply them.

It seems that paths are alphabetically ordered right now, which means for an array larger than 10 items, an operation for e.g. index 11 happen is attempted before an operation for e.g. index 2.

Further, for an array of size 3 e.g. a `replace` operation on index 2 might fail if there is a `remove` operation for index 1.

It looks like the whole array would need to be replaced in order to apply the changes cleanly.

This is a blocker for 1.0.

**Comment by michaelsauter** [comment]
Closed by #168.
